tests/callbacks/test_device_stats_monitor.py

```diff
@@ -85,6 +102,100 @@ def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) ->
+    device_stats = DeviceStatsMonitor(cpu_stats=cpu_stats)
+
+    class DebugLogger(CSVLogger):
+        @rank_zero_only
+        def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
+            # Just need to check one of the GPU metrics to make sure
+            # the GPU metrics are logged
+            fields = [
+                "active.all.allocated",
+            ]
+
+            # If cpu stats, also check CPU metric keys are logged
+            fields += CPU_METRIC_KEYS if cpu_stats else []
+            for f in fields:
+                assert any(f in h for h in metrics)
+
+            # If not cpu stats, make sure CPU metric keys aren't logged
+            if not cpu_stats:
+                assert not any(f in h for h in metrics), "CPU Stats should not be included"
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=2,
+        log_every_n_steps=1,
+        accelerator="gpu",
+        devices=1,
+        callbacks=[device_stats],
+        logger=DebugLogger(tmpdir),
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+
+    trainer.fit(model)
+
+
+def test_device_stats_cpu(tmpdir):
+    """Test CPU stats are logged when no accelerator is used."""
+    model = BoringModel()
+
+    class DebugLogger(CSVLogger):
+        @rank_zero_only
+        def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
+            fields = CPU_METRIC_KEYS
+            for f in fields:
+                assert any(f in h for h in metrics)
+
+    device_stats = DeviceStatsMonitor()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=2,
+        limit_train_batches=7,
+        log_every_n_steps=1,
+        callbacks=[device_stats],
+        logger=DebugLogger(tmpdir),
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+
+    trainer.fit(model)
+
+
+@mock.patch("pytorch_lightning.accelerators.cpu.get_cpu_process_metrics")
+@mock.patch("pytorch_lightning.callbacks.device_stats_monitor.get_cpu_process_metrics")
+def test_device_stats_cpu_queried_once(cpu_metrics_device_stats_mock, cpu_metrics_cpu_accelerator_mock, tmpdir):
+    """Make sure that get_cpu_process_metrics is only queried once if the accelerator is CPU and cpu_stats=True."""
+    model = BoringModel()
+
+    device_stats = DeviceStatsMonitor(cpu_stats=True)
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_train_batches=1,
+        log_every_n_steps=1,
+        callbacks=[device_stats],
+        logger=True,
+        enable_checkpointing=False,
+        enable_progress_bar=False,
+    )
+
+    trainer.fit(model)
+
+    # Note that you need to mock where the function is imported
+    # (not where it is defined). Please see the following for
+    # an explanation: https://docs.python.org/3/library/unittest.mock.html#where-to-patch
```

Comment:
While I see that error made often by people, let's remove this comment. This should be known by everybody who uses unittest mocks, and if not, a comment in a random test is not the place to educate :)